sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/TableHeaderHasIdOrScopeCheck.java

```diff
@@ -17,22 +17,100 @@
 public class TableHeaderHasIdOrScopeCheck extends AbstractPageCheck {
 
+  private static final String MESSAGE = "Add either an 'id' or a 'scope' attribute to this <th> tag.";
+
+  private Deque<TableElement> tables = new ArrayDeque<>();
+
   @Override
   public void startElement(TagNode node) {
-    if (isTableHeader(node) && !node.hasProperty("ID") && !node.hasProperty("SCOPE")) {
-      createViolation(node, "Add either an 'id' or a 'scope' attribute to this <" + node.getNodeName() + "> tag.");
+    if (isTableTag(node)) {
+      tables.push(new TableElement());
+    }
+
+    if (isTrTag(node)) {
+      visitTrNode(node);
     }
   }
 
-  private static boolean isTableHeader(TagNode node) {
-    return "TH".equalsIgnoreCase(node.getNodeName());
+  private void visitTrNode(TagNode node) {
+    List<TagNode> row = node.getChildren();
+    if (!tables.isEmpty() && !row.isEmpty()) {
+
+      TableElement currentTable = tables.peek();
+      for (TagNode element : row) {
+        if (isThTag(element)) {
+          currentTable.headers.add(element);
+        }
+      }
+      currentTable.firstCol.add(row.get(0));
+
+      if (currentTable.firstRow.isEmpty()) {
+        currentTable.firstRow = row;
+      }
+
+    } else {
+      // Sometimes rows are defined in separate files, and at this point we don't see them as part of a current table.
+      // In this case we treat them as belonging to "not simple tables".
+      raiseIssueOnTableHeadersWithoutScopeOrId(row);
+    }
+  }
+
+  @Override
+  public void endElement(TagNode node) {
+    if (isTableTag(node) && !tables.isEmpty()) {
+      TableElement currentTable = tables.peek();
+      if (!currentTable.isSimpleTable()) {
+        raiseIssueOnTableHeadersWithoutScopeOrId(currentTable.headers);
+      }
+      tables.pop();
+    }
   }
 
+  private static boolean isThTag(TagNode node) {
+    return "th".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean isTrTag(TagNode node) {
+    return "tr".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private static boolean isTableTag(TagNode node) {
+    return "table".equalsIgnoreCase(node.getNodeName());
+  }
+
+  private void raiseIssueOnTableHeadersWithoutScopeOrId(Collection<TagNode> row) {
+    row.stream()
+      .filter(TableHeaderHasIdOrScopeCheck::isHeaderTableWithoutScopeOrId)
+      .forEach(th -> createViolation(th, MESSAGE));
+  }
+
+  private static boolean isHeaderTableWithoutScopeOrId(TagNode child) {
+    return isThTag(child) && !child.hasProperty("id") && !child.hasProperty("scope");
+  }
+
+  private static class TableElement {
+    List<TagNode> headers = new ArrayList<>();
+    List<TagNode> firstRow = new ArrayList<>();
+    List<TagNode> firstCol = new ArrayList<>();
+
+    /**
+     * We consider as simple tables, tables which have all headers only in the first row or in the first column.
+     * If both first row and first column are all (or partially) composed by headers are not considered simple tables.
```

Comment:
I'm not sure I understand the second sentence. Did you mean to say something like "If both first row and first column are all (or partially) composed by headers, the table is not considered a simple tables."

Reply:
yes, but since it is not clear I will update this. Does "We consider as simple tables, tables with the first row or the first column only composed by headers.  The two conditions must not apply together." sounds clearer?

Reply:
Wait, this is ambiguous.. I will reformulate :D

Reply:
"Simple tables are considered as such when the headers are either all in the first row, or all in the first column. The two conditions must not apply together."